KsApi/PKCE.swift

```diff
@@ -0,0 +1,58 @@
+import CryptoKit
+import Foundation
+
+/// PKCE stands for Proof Key for Code Exchange.
+/// The code verifier, and its associated challenge, are used to ensure that an oAuth request is valid.
+/// See this documentation for more details: https://www.oauth.com/oauth2-servers/mobile-and-native-apps/authorization/
+public struct PKCE {
+  /// Creates a random alphanumeric string of the specified length
+  static func createCodeVerifier(ofLength length: Int) throws -> String {
+    let uppercase = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
+    let lowercase = "abcdefghijklmnopqrstuvwxyz"
+    let numbers = "0123456789"
+    let specials = "-._~"
+
+    let validCharacters = uppercase + lowercase + numbers + specials
+
+    var codeVerifier = ""
+
+    for _ in 0..<length {
+      guard let character = validCharacters.randomElement() else {
```

Comment:
@Arkariang So this is something I wanted to verify - `randomElement()` uses the system random number generator. Is that going to be _good enough_ for PKCE? The documentation says "cryptographically random" and it's not clear if it needs a more random kind of random than the system provides.

Reply:
I hesitate to rely on the default random generator for crypto - all my crypto classes in college taught "never implement crypto yourself if you can avoid it". However, I did a little digging and it looks fairly safe.

Reasoning: https://developer.apple.com/documentation/swift/systemrandomnumbergenerator claims that the system default random number generator is cryptographically secure whenever possible. https://developer.apple.com/documentation/security/randomization_services, which is cryptographically secure, uses the system default as its source of randomness (unsure what it does if the system default isn't good enough). `.randomElement` uses the system default as well.

The main problem with this approach is that I'm not sure what happens if the system default isn't secure enough (which could happen if there isn't a lot of entropy to draw from - ex new computers booted up at the same time will probably have the same seed and get the same "random" numbers). It's possible that Apple's cryptographically secure random bytes method would throw an error instead in this case, and that that's what makes it actually cryptographically random. However, I don't know enough about PKCE to know if this would actually be a problem we need to handle; I expect our real users would have cryptographically random verifiers.

I think I'd recommend leaving this as-is for now and going to security office hours next week. Or looking for maybe 3rd party solutions for getting guaranteed cryptographic randomness. (Or using https://developer.apple.com/documentation/security/1399291-secrandomcopybytes, but then we can't specify what kinds of bytes we're okay with, as far as I can tell.)

Reply:
I think it is refered to use some kind of secure hash algorithm, in my case for Android I'll use SHA-256